Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs

```diff
@@ -239,5 +194,69 @@ private IEnumerable<Type> GetChildTypes(Type type)
 
             return allChildTypes[type];
         }
+
+        private IEnumerable<CandidateChildType> GetCandidateChildTypes(Type objectType, HashSet<string> jsonProperties)
+        {
+            var childTypes = GetChildTypes(objectType);
+            var candidateChildTypes = new List<CandidateChildType>();
+            foreach (var childType in childTypes)
+            {
+                var childTypeProperties = childType.GetProperties().Select(t => t.Name).ToList();
+                var childTypeFields = childType.GetFields().Select(t => t.Name).ToList();
+                var childTypePropertiesAndFields = new HashSet<string>(
+                    childTypeProperties.Concat(childTypeFields),
+                    StringComparer.OrdinalIgnoreCase);
+                if (jsonProperties.All(p => childTypePropertiesAndFields.Contains(p)))
+                {
+                    var candidateChildType = new CandidateChildType
+                    {
+                        Type = childType,
+                        PropertiesAndFields = childTypePropertiesAndFields
+                    };
+                    candidateChildTypes.Add(candidateChildType);
+                }
+            }
+
+            return candidateChildTypes;
+        }
+
+        private void DeserializeCandidates(JsonSerializer serializer, JObject jsonObject, IEnumerable<CandidateChildType> candidateChildTypes)
+        {
+            candidateChildTypes = candidateChildTypes.ToList();
+            foreach (var candidateChildType in candidateChildTypes)
+            {
+                try
+                {
+                    var deserializedObject = serializer.Deserialize(jsonObject.CreateReader(), candidateChildType.Type);
+                    if (deserializedObject != null)
+                    {
+                        candidateChildType.DeserializedObject = deserializedObject;
```

Comment:
I wasn't referring to technically making the object immutable. However, that doesn't mean you still shouldn't treat it as such. You're right, C# doesn't give us a clean mechanism for declaring we want our types to be immutable without a bunch of constructor noise. I tend to ignore that noise, use POCOs, and still treat those types as if they were immutable. We can still be immutable at runtime without the compiler guarantee if we simply never modify the objects outside of creation time.